include/deal.II/matrix_free/operators.h

```diff
@@ -30,6 +30,224 @@ DEAL_II_NAMESPACE_OPEN
+
+    /**
+     * Initialize operator on a level @p level.
+     */
+    void initialize (const MatrixFree<dim,number> &data,
+                     const MGConstrainedDoFs &mg_constrained_dofs,
+                     const unsigned int level);
+
+    /**
+     * Return the dimension of the codomain (or range) space.
+     */
+    unsigned int m () const;
+
+    /**
+     * Return the dimension of the domain space.
+     */
+    unsigned int n () const;
+
+    /**
+     * vmult operator for interface.
+     */
+    void vmult_interface_down(LinearAlgebra::distributed::Vector<number> &dst,
+                              const LinearAlgebra::distributed::Vector<number> &src) const;
+
+    /**
+     * vmult operator for interface.
+     */
+    void vmult_interface_up(LinearAlgebra::distributed::Vector<number> &dst,
+                            const LinearAlgebra::distributed::Vector<number> &src) const;
+
+    /**
+     * Tvmult operator for interface.
+     */
+    void Tvmult_interface_down(LinearAlgebra::distributed::Vector<number> &dst,
+                               const LinearAlgebra::distributed::Vector<number> &src) const;
+
+    /**
+     * Tvmult operator for interface.
+     */
+    void Tvmult_interface_up(LinearAlgebra::distributed::Vector<number> &dst,
+                             const LinearAlgebra::distributed::Vector<number> &src) const;
+
+
+    /**
+     * Matrix-vector multiplication.
+     */
+    void vmult (LinearAlgebra::distributed::Vector<number> &dst,
+                const LinearAlgebra::distributed::Vector<number> &src) const;
+
+    /**
+     * Transpose matrix-vector multiplication.
+     */
+    void Tvmult (LinearAlgebra::distributed::Vector<number> &dst,
+                 const LinearAlgebra::distributed::Vector<number> &src) const;
+
+    /**
+     * Adding Matrix-vector multiplication.
+     */
+    void vmult_add (LinearAlgebra::distributed::Vector<number> &dst,
+                    const LinearAlgebra::distributed::Vector<number> &src) const;
+
+    /**
+     * Adding transpose matrix-vector multiplication.
+     */
+    void Tvmult_add (LinearAlgebra::distributed::Vector<number> &dst,
+                     const LinearAlgebra::distributed::Vector<number> &src) const;
+
+    /**
+     * Returns the value of the matrix entry (row,col). In matrix-free context
+     * this function is valid only for row==col when diagonal is initialized.
+     */
+    number el (const unsigned int row,
+               const unsigned int col) const;
+
+    /**
+     * Determine an estimate for the memory consumption (in bytes) of this object.
+     */
+    virtual std::size_t memory_consumption () const;
+
+    /**
+     * A wrapper for initialize_dof_vector() of MatrixFree object.
+     */
+    void initialize_dof_vector (LinearAlgebra::distributed::Vector<number> &vec) const;
+
+    /**
+     * Compute diagonal of this operator. To be implemented in derived classes.
+     */
+    virtual void compute_diagonal () = 0;
```

Comment:
I would keep it as pure virtual which would make users to at least add a dummy implementation with `AssertThrow(false, ExcNotImplemented())`.